packages/components/src/lib/controllers/ui-controller.js

```diff
@@ -891,6 +905,31 @@ if you want to create a different configuration of a UI controller.
     return [this.featureOptions.load(), this.resourceOptions.load(), this.uiOptions.load()]
   }
 
+  /**
+   * Loads an application wide configuration file in a JSON format.
+   *
+   * @param {string} url - A URL of an app config server.
+   * @param {string} clientId - A client ID.
+   * @param {string} appName - An application name, such as "Alpheios Reading Tools".
+   * @param {string} appVersion - An application version, following a semver specification.
+   * @param {string} branch - A name of a git branch that was used to build the code.
+   * @param {string} buildNumber - A build number in a YYYYMMDDCCC format.
+   *        See `node-build` package for a generator function.
+   * @returns {Promise<any>} - A promise that is resolved with an app config in a JSON format
+   *          or is rejected with an error if a config retrieval failed.
+   */
+  async loadAppConfig ({ url, clientId, appName, appVersion, branch, buildNumber } = {}) {
+    if (!url) {
+      throw new Error('An app config server URL is missing')
+    }
+    const configUrl = `${url}?clientId=${encodeURIComponent(clientId)}&appName=${encodeURIComponent(appName)}` +
+      `&appVersion=${encodeURIComponent(appVersion)}&buildBranch=${encodeURIComponent(branch)}` +
+      `&buildNumber=${encodeURIComponent(buildNumber)}`
+    const request = new Request(configUrl)
+    const response = await fetch(request)
```

Comment:
what happens if the request fails?

Reply:
and in the unit tests, will the fetch call be executed or mocked?

Reply:
> what happens if the request fails?

If a request failed, fetch will throw a "Failed to fetch" exception that will fall all the way through. As a result, an embed-lib or webextension will get that exception when they will try to activate a UI controller. I think that's what's desired because thee is no fallback if the remote config server is down. What do you think?

It may only be better to catch a "Failed to fetch" and rethrow as something saying that a remote config cannot be loaded. Would it make it more understandable?

Reply:
> and in the unit tests, will the fetch call be executed or mocked?

Unfortunately, we have a majority of UI controller tests commented out after refactoring. As a result of this, this is not tested currently. I can work on enabling those tests but that will require a serious time and effort.

Reply:
ok. I will add an issue for that.

Does the way this is coded prevent that call from being able to be mocked?

Reply:
> > what happens if the request fails?
> 
> If a request failed, fetch will throw a "Failed to fetch" exception that will fall all the way through. As a result, an embed-lib or webextension will get that exception when they will try to activate a UI controller. I think that's what's desired because thee is no fallback if the remote config server is down. What do you think?
> 
> It may only be better to catch a "Failed to fetch" and rethrow as something saying that a remote config cannot be loaded. Would it make it more understandable?

I think we should not prevent the UIController from loading if the config fails to load. It should just disable those features that are dependent upon it.

Reply:
> Does the way this is coded prevent that call from being able to be mocked?

Nope. There is a separate function in a UI controller that takes request parameters and returns a config JSON. It could be replaced with a mock easily.

Reply:
> I think we should not prevent the UIController from loading if the config fails to load. It should just disable those features that are dependent upon it.

OK, I will change code to handle that.

Reply:
ok thanks!